lib/rules/quote-props.js

```diff
@@ -68,5 +97,14 @@ module.exports = function(context) {
 module.exports.schema = [
     {
         "enum": ["always", "as-needed"]
+    },
+    {
+        "type": "object",
```

Comment:
First, I want to check my understanding of how this is supposed to work.

```js
{
    "quote-props": 2, // Valid, uses defaults
    "quote-props": [2], // Valid, same as above
    "quote-props": [2, "always"], // Valid, `keywords` does not apply
    "quote-props": [2, "as-needed"], // Valid, uses default value for `keywords`
    "quote-props": [2, "as-needed", { "keywords": true }], // Valid, specifies value for `keywords`
    "quote-props": [2, "always", { "keywords": true }] // Invalid, `keywords` cannot be used with `always`
}
```

Am I correct on that? Now would be a good time to read the [documentation on schemas](http://eslint.org/docs/developer-guide/working-with-rules#options-schemas).

Almost all of our rules don't have inter-property dependencies. That means we can pass an array of individual schemas for each configuration value, and the validator can handle each value separately. Recall from the docs that there are two ways to pass a schema, the array being the more convenient and encouraged method. In the case of [no-restricted-modules](https://github.com/eslint/eslint/blob/master/lib/rules/no-restricted-modules.js#L74-L85), the previous assumptions don't apply, so we have to pass a full schema. We'll need to do the same here:

```js
module.exports.schema = {
    "anyOf": [
        {
            "type": "array",
            "items": [
                {
                    "enum": [0, 1, 2]
                },
                {
                    "enum": ["always"]
                }
            ],
            "minItems": 1,
            "maxItems": 2
        },
        {
            "type": "array",
            "items": [
                {
                    "enum": [0, 1, 2]
                },
                {
                    "enum": ["as-needed"]
                },
                {
                    "type": "object",
                    "properties": {
                        "keywords": {
                            "type": "boolean"
                        }
                    },
                    "additionalProperties": false
                }
            ],
            "minItems": 1,
            "maxItems": 3
        }
    ]
};
```

The previous schema provides two possible matching alternatives, and the provided options must match at least one of them. Both alternatives require a warning level. The first allows for an optional "always" mode. The second allows for an optional "as-needed" mode and an optional "keywords" option.